source/common/network/udp_listener_impl.cc

```diff
@@ -167,6 +167,7 @@ void UdpListenerImpl::send(const UdpSendData& send_data) {
       socket_.ioHandle().sendmsg(slices.begin(), num_slices, 0, *send_data.send_address_);
 
   if (send_result.ok()) {
+    buffer.drain(send_result.rc_);
```

Comment:
nit: move buffer.drain() outside of if...else... block as it needs to be called in both cases. Add `ASSERT(send_result.rc_ == buffer.length())` here.

Reply:
I can add the assert. 

I don't follow the other comment: on error, -1 is the value of rc_. How can drain be called with -1 ? The parameter type is uin64 in drain - do you intend we should call this with -1 that gets converted to max int64? what is the behavior if we try to drain more than what is in the buffer?

Reply:
In both cases we drain the whole buffer, right?

Reply:
No, from my conversation with @mattklein123 earlier, only bytes successfuly sent are to be drained. The remaining are left for the caller to decide if they want to retry or chunk up the contents if the buffer was too big.

Reply:
If there are remaining bytes, it should return Again instead of OK right?

Reply:
That's right. do you suggest we check for the "-1" and only drain if the rc_ is positive?

Reply:
I see your point: I missed the fact that the -1 from sendmsg is getting converted in the os_sys_calls layer and surfaced as 0.

So yes, we can invoke drain outside the if else loop.

Reply:
I couldn't find the previous discussion about the intended behavior. I thought it was a best-effort send so draining the whole buffer regardless of the result makes sense. But if we want to leave the door  open to continue sending, a more sophisticated send is needed, i.e. sendmsg can also failure upon EINTR. Would it make more sense to simply return the result and leave the decision to the caller whether to drain the buffer or not?

If the previous decision was to drain the buffer here. I think drain the whole buffer upon success and errors other than Again makes sense. Returning Again means we didn't send any bytes, it's upto caller to handle the buffer.

Reply:
Here is an example of a sophisticated UDP send behavior regarding EINTR:
https://cs.chromium.org/chromium/src/net/quic/platform/impl/quic_socket_utils.cc?rcl=84dbdd788a32c2ae1ede0ed60a47c45cd3553d14&l=331

Reply:
Thanks for the pointer - I tagged you in the previous conversation about the drain behavior. Let's try to get this resolved.